### DISCUSSION TITLE
how to simplify AST?

### DISCUSSION THREAD

**Discussion by shi-yan** [tree-sitter/tree-sitter#1447] [Q&A]
Hello,

say I'm trying to implement an interpreter for an existing language. I'd like to use tree-sitter as the frontend.

what I found out is that the AST output of tree-sitter could be very verbose, for example,
```cpp
enum class AAA {
    BBB,
}
```

could generate
```
(enum_definition
(enum_keyword)
(class_keyword)
(identifier AAA)
(enum_item_list
    (left_bracket)
    (enum_item_list_body
        (enum_item
            (identifier BBB)
        )
    )
    (right_bracket)
)

```

from the interpreter's  perspective, perhaps I all I need is this

```
(enum_definition AAA 
[(enum_item BBB), ....]
)
```

How can I simplify the verbose AST into the simple version? Sorry this is not a question about tree-sitter, but a general question regarding interpreter implementation. I don't know the technical term for this simplification (so that I can google more)? is it term/tree rewriting ?

what's the name for the simplified AST?  like for example (Parse Tree/concrete syntax tree) -> AST -> ?

And I don't know what's the most effective way to create code that can simplify AST, because it's difficult to cover every corner of the targeted language.

**Comment by cfroystad** [comment]
This is not a complete answer, but hopefully a pointer in the right direction...

For your specific question on simplifying the output relevant to your need, you might find some relevant articles by searching for "reducing cst to ast" or variations thereof.

However, if you're planning to do this at scale, it might (I've neither tested nor thought deeply about this) be better to [walk the tree](https://tree-sitter.github.io/tree-sitter/using-parsers#walking-trees-with-tree-cursors) and build what you need from the encountered nodes.
